### PR TITLE
Feature/print

### PR DESCRIPTION
**Pull request by apmccartney**
+ added `NC` getters to most types
+ added `print` function to most types
+ refactored the unresolved resonance classes
+ renamed the `UnivariateTabulation` to `TabulationRecord`
+ added `Lvalue` and `Jvalue`/`Resonance` classes to resonance types
+ resonance types provide `LRO` and `LRF` getters

### CODE REVIEW

File: src/ENDFtk/ControlRecord.hpp [hunk 1]
```diff
@@ -54,4 +54,8 @@ class ControlRecord : protected record::Base< record::Real,
   operator!=( const ControlRecord& rhs ) const {
     return not( *this == rhs );
   }
+
+  int NC() const { return 1; }
```
Comment (jlconlin) [inline src/ENDFtk/ControlRecord.hpp]: What is `NC`?

File: src/ENDFtk/DirectoryRecord.hpp [hunk 2]
```diff
@@ -48,5 +48,9 @@ class DirectoryRecord : protected record::Base< record::Integer< 33 >,
   bool
   operator!=( const DirectoryRecord& rhs ) const {
     return not( *this == rhs );
-  }  
+  }
+
+  int NC() const { return 1; }
```
Comment (jlconlin) [inline src/ENDFtk/DirectoryRecord.hpp]: What is `NC`?

File: src/ENDFtk/ControlRecord.hpp (hunk 1 above)
Comment (jlconlin) [reply src/ENDFtk/ControlRecord.hpp]: Oh, I get it now. It's the number of lines in the ENDF file.

File: src/ENDFtk/resonanceParameters/resolved/BreitWigner/Lvalue.hpp [hunk 3]
```diff
@@ -0,0 +1,27 @@
+class Lvalue {
```
Comment (jlconlin) [inline src/ENDFtk/resonanceParameters/resolved/BreitWigner/Lvalue.hpp]: I'm not real keen on calling this `Lvalue` instead of `LState`. To me an l-value is an integer quantity specifying an independent. Whereas the l-state is the l-value plus the rest of the information describing the physics.

File: src/ENDFtk/resonanceParameters/unresolved/EnergyDependent/Jvalue.hpp [hunk 4]
```diff
@@ -0,0 +1,29 @@
+class Jvalue {
```
Comment (jlconlin) [inline src/ENDFtk/resonanceParameters/unresolved/EnergyDependent/Jvalue.hpp]: For the same reason I don't like the name `Lvalue`, I also don't like the name `Jvalue`.

Comment (jlconlin) [reply src/ENDFtk/resonanceParameters/unresolved/EnergyDependent/Jvalue.hpp]: Also, shouldn't the `Jvalue`/`JState` be a class inside `LState`?

Comment (apmccartney) [reply src/ENDFtk/resonanceParameters/unresolved/EnergyDependent/Jvalue.hpp]: <yoda> Do or do not; there is no should </yoda>

What motivates the preference?

File: src/ENDFtk/resonanceParameters/resolved/BreitWigner/Lvalue.hpp (hunk 3 above)
Comment (apmccartney) [reply src/ENDFtk/resonanceParameters/resolved/BreitWigner/Lvalue.hpp]: I wasn't fond of either name. I checked openMC and DECE to see if there were other more respresentative or descriptive names for this aggregate, but they both seem to work in terms of the primitives without provides this abstraction.

That said, I favor `Lvalue` as it mirrors the naming in the resonance reconstruction repo and refers to a physical quantity domain experts will recognize, as opposed to `LState`, which is a term we invented and is less likely to have an immediate association

**Comment by apmccartney** [comment]
`NC` comments resolved and deleted

File: src/ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths/src/readLvalues.hpp [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  if ( NLS < 1 ){
+    Log::error( "NLS required to be positive" );
+    Log::error( "NLS value: {}", NLS );
+    throw std::exception();
+  }
+  
+  lValues.reserve( NLS );
+  while ( NLS-- ){ lValues.emplace_back( it, end, lineNumber, MAT, MF, MT ); }
+  return lValues;
+}
```
Comment (whaeck) [inline src/ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths/src/readLvalues.hpp]: This function is implemented multiple times, using a different Lvalue class. Maybe add this as a template function in a base class?

File: src/ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths/test/EnergyDependentFissionWidths.test.cpp [hunk 6]
```diff
@@ -63,8 +60,8 @@ SCENARIO( "Testing energy-dependent fission width unresolved resonances" ){
           base, begin, end, lineNumber, MAT, MF, MT );
 
       THEN( "the parameters can be verified" ){
-        REQUIRE( 5.7E3 == edfw.EL() );
-        REQUIRE( 4.0E4 == edfw.EH() );
+        REQUIRE( Approx(5.7E3) == edfw.EL() );
+        REQUIRE( Approx(4.0E4) == edfw.EH() );
```
Comment (whaeck) [inline src/ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths/test/EnergyDependentFissionWidths.test.cpp]: How come you did not have a roundoff issue before?

File: src/ENDFtk/resonanceParameters/unresolved/EnergyIndependent.hpp [hunk 7]
```diff
@@ -8,15 +12,17 @@ class EnergyIndependent {
...
-
+  
 public:
+  #include "ENDFtk/resonanceParameters/unresolved/EnergyIndependent/src/ctor.hpp"
+  
   double EL() const { return this->el; }
   double EH() const { return this->eh; }
-  int NRO() const { return bool(this->ape); }
+  int NRO() const { return bool( this->ape ); }
   int NAPS() const { return this->naps; }
```
Comment (whaeck) [inline src/ENDFtk/resonanceParameters/unresolved/EnergyIndependent.hpp]: As above.

File: src/ENDFtk/resonanceParameters/unresolved/EnergyIndependent.hpp [hunk 8]
```diff
@@ -25,7 +31,21 @@ class EnergyIndependent {
   int LSSF() const { return this->lssf; }
   int NLS() const { return this->lists.size(); }
```
Comment (whaeck) [inline src/ENDFtk/resonanceParameters/unresolved/EnergyIndependent.hpp]: As above: LRU, LRF, LSSF, NLS

File: src/ENDFtk/resonanceParameters/unresolved/EnergyIndependent.hpp [hunk 9]
```diff
@@ -25,7 +31,21 @@ class EnergyIndependent {
...
-  auto LStates() const {
-    return ranges::make_iterator_range( this->lists.begin(), this->lists.end() );
+  auto lValues() const {
+    return
+      this->lists
+      | ranges::view::transform
+        ( []( const auto& list ){ return Lvalue( list ); } );
   }
+
+  int NC() const {
```
Comment (whaeck) [inline src/ENDFtk/resonanceParameters/unresolved/EnergyIndependent.hpp]: Same as above.

File: src/ENDFtk/resonanceParameters/unresolved/EnergyIndependent/Lvalue.hpp [hunk 10]
```diff
@@ -0,0 +1,17 @@
+class Lvalue {
+  const LIST& core;
+
+public:
+  Lvalue( const LIST& list ) : core( list ){}
+
+  double AWRI() const { return this->core.C1(); }
+  int L() const { return this->core.L1(); }
+  int NJS() const { return this->core.N2(); }
```
Comment (whaeck) [inline src/ENDFtk/resonanceParameters/unresolved/EnergyIndependent/Lvalue.hpp]: Same as above.

**Comment by jlconlin** [comment]
This is a good place for nitpicking. We can discuss why we want things one way vs another. And having it on GitHub makes it possible to keep a record of our decisions (if only we could search them.)

File: src/ENDFtk/ControlRecord/test/ControlRecord.test.cpp [hunk 11]
```diff
@@ -7,7 +7,7 @@ using namespace njoy::ENDFtk;
 
 SCENARIO( "ControlRecord Tests", "[ENDFtk], [ControlRecord]" ){
   std::string line =
-    " 1.001000+3 9.991673-1          0          0          0          5 125 1451    1\n";
+    " 1.001000+3 9.991673-1          0          0          0          5 125 1451     \n";
```
[the remark this answers is not in the document]
Comment (apmccartney) [reply src/ENDFtk/ControlRecord/test/ControlRecord.test.cpp]: This functionality would require a running total for `NC`, which could be supplied as an argument as MAT, MF, and MT are; easy to accomodate with an overload.

That said, until someone requests the functionality, I'm happy to defer that work, especially in light of the deadline at the end of the month.

File: src/ENDFtk/ListRecord/src/print.hpp [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+                                  disco::ENDF, disco::ENDF,
+                                  disco::Integer< 4 >, disco::Integer< 2 >,
+                                  disco::Integer< 3 >, disco::ColumnPosition<5> >;
+
+    auto nFullLines = this->NPL() / 6;
+    auto partialLineEntries = this->NPL() - nFullLines * 6;
+    
+    auto entry = this->list().begin();
+
+    for ( ; nFullLines--; entry += 6 ){
```
[the remark this answers is not in the document]
Comment (apmccartney) [reply src/ENDFtk/ListRecord/src/print.hpp]: For random access iterators, `entry += 6` and `std::advance(entry, 6)` are equivalent.
For less powerful iterator categories, there is no `+=` overload and `std::advance` increments `n` times.

File: src/ENDFtk/ControlRecord/test/ControlRecord.test.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (whaeck) [reply src/ENDFtk/ControlRecord/test/ControlRecord.test.cpp]: I concur. No need to anticipate a user's request that adds little to the toolkit.

File: src/ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths/test/EnergyDependentFissionWidths.test.cpp (hunk 6 above)
Comment (apmccartney) [reply src/ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths/test/EnergyDependentFissionWidths.test.cpp]: On the platforms we test on, we don't have round-off then or now. That said, floating point equality is squirrelly. 

Just a good habit when working with floating point numbers.

**Comment by apmccartney** [comment]
Given the suggested number of changes from `int` to `long`, it appears I'm pretty consistent in using `int` values.

File: src/ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths/src/readLvalues.hpp (hunk 5 above)
Comment (apmccartney) [reply src/ENDFtk/resonanceParameters/unresolved/EnergyDependentFissionWidths/src/readLvalues.hpp]: No need for a base class for what would be a static function, but point taken

File: src/ENDFtk/ControlRecord.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (apmccartney) [reply src/ENDFtk/ControlRecord.hpp]: The value-to-effort ratio is pretty low for these requests and it pessimizes storage on the client side when they use auto.

File: src/ENDFtk/ListRecord/src/print.hpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (apmccartney) [reply src/ENDFtk/ListRecord/src/print.hpp]:
```cpp
nFullLines-- != 0
```

Eh, I think its noisy, but I'm probably in the minority so I'll give you this one

File: src/ENDFtk/ControlRecord.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (whaeck) [reply src/ENDFtk/ControlRecord.hpp]: I understand that. I would argue that if we use long (or any other type) as the type for integer values stored in an 11 character column of the ENDF file (meaning L1, L2, N1, N2, interpolation types and indices), any ENDF defined parameter should use the same type (even if it is not stored in a record). Hence the remark that NC, LRF, etc. should be long (again, if we use long as the storage type).
However, the largest integer value that I have ever seen in an ENDF evaluation is about 250000 (which is NC for ENDF/B-VII.1 U235 MF1 MT460). Some covariance sections also reach NC values of this order of magnitude. Larger integer values could occur in pointwise reconstructed cross section NP values (but I would be surprised that this would be larger than say 1000000 unless of course you use an outrageously small reconstruction tolerance). Taking this in consideration, I agree that int would be more than sufficient.

File: src/ENDFtk/resonanceParameters/src/read.hpp [hunk 13]
```diff
@@ -1,8 +1,8 @@
-template< typename Iterator >
+template< typename Lvalue, typename Iterator >
 static std::vector< Lvalue >
-readLvalues( int NLS,
-             Iterator& it, const Iterator& end, long& lineNumber,
-             int MAT, int MF, int MT ) {
+read( int NLS,
```
Comment (jlconlin) [inline src/ENDFtk/resonanceParameters/src/read.hpp]: Why did you change the name of the function?

File: src/ENDFtk/ControlRecord.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (apmccartney) [reply src/ENDFtk/ControlRecord.hpp]: > I  would argue that if we use long (or any other type) as the type for integer values stored in an 11 character column of the ENDF file (meaning L1, L2, N1, N2, interpolation types and indices), any ENDF defined parameter should use the same type (even if it is not stored in a record). 

why though

File: src/ENDFtk/resonanceParameters/src/read.hpp (hunk 13 above)
Comment (apmccartney) [reply src/ENDFtk/resonanceParameters/src/read.hpp]: I didn't. New function similar enough to the previous, git thought it was a moved file

File: src/ENDFtk/ControlRecord.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (apmccartney) [reply src/ENDFtk/ControlRecord.hpp]: despite being an 11 character field, `LRU`, for example, can only assume 3 values: 0, 1, 2.

File: src/ENDFtk/resonanceParameters/src/read.hpp (hunk 13 above)
Comment (jlconlin) [reply src/ENDFtk/resonanceParameters/src/read.hpp]: The function used to be called `readLvalues`, now it is called `read`—regardless of what git does.

File: src/ENDFtk/ControlRecord.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (whaeck) [reply src/ENDFtk/ControlRecord.hpp]: I know. It's just that I like uniformity and consistency over the entire toolkit :-)
If you do it this way for MF2 MT451, I will do it this way for everything I put in (starting with file 1).
